File: frontend/lib/shared/helpers/nutrition_goal_calculator.dart
class NutritionGoalInput {
  const NutritionGoalInput({
    required this.weight,
    required this.height,
    required this.age,
    required this.sex,
    required this.objective,
    required this.activityLevel,
    this.weightUnit = 'kg',
    this.heightUnit = 'cm',
  });

  final double weight;
  final double height;
  final int age;
  final String sex;
  final String objective;
  final String activityLevel;
  final String weightUnit;
  final String heightUnit;
}

class NutritionGoalResult {
  const NutritionGoalResult({
    required this.effectiveWeightKg,
    required this.effectiveHeightCm,
    required this.dailyCalorieGoal,
    required this.dailyProteinGoal,
    required this.dailyCarbsGoal,
    required this.dailyFatGoal,
  });

  final double effectiveWeightKg;
  final double effectiveHeightCm;
  final int dailyCalorieGoal;
  final int dailyProteinGoal;
  final int dailyCarbsGoal;
  final int dailyFatGoal;
}

NutritionGoalResult calculateNutritionGoals(NutritionGoalInput input) {
  final weightKg = _toKilograms(input.weight, input.weightUnit);
  final heightCm = _toCentimeters(input.height, input.heightUnit);

  final age = input.age.clamp(0, 150).toInt();
  final objective = _normalizeObjective(input.objective);
  final bmi = _calculateBmi(weightKg, heightCm);
  final metabolicWeightKg = _resolveMetabolicWeightKg(
    weightKg,
    heightCm,
    objective,
  );

  final baseMetabolism = input.sex == 'Feminino'
      ? (10 * metabolicWeightKg) + (6.25 * heightCm) - (5 * age) - 161
      : (10 * metabolicWeightKg) + (6.25 * heightCm) - (5 * age) + 5;

  final activityMultiplier = switch (input.activityLevel) {
    'lightly' => 1.375,
    'moderate' => 1.55,
    'very' => 1.725,
    'extreme' => 1.9,
    _ => 1.2,
  };

  final tdee = baseMetabolism * activityMultiplier;
  final adjustedCalories = _applyObjectiveAdjustment(
    tdee,
    objective,
    bmi,
    input.activityLevel,
  );
  final minimumCalories = _resolveCalorieFloor(input.sex);
  final dailyCalorieGoal = adjustedCalories.round().clamp(minimumCalories, 9999);

  final macros = _resolveMacroGoals(
    dailyCalorieGoal,
    objective,
    metabolicWeightKg,
  );

  return NutritionGoalResult(
    effectiveWeightKg: metabolicWeightKg,
    effectiveHeightCm: heightCm,
    dailyCalorieGoal: dailyCalorieGoal,
    dailyProteinGoal: macros.protein,
    dailyCarbsGoal: macros.carbs,
    dailyFatGoal: macros.fat,
  );
}

int calculateAgeFromBirthDate(String? value, {DateTime? referenceDate}) {
  final birthDate = _parseBirthDate(value);
  if (birthDate == null) {
    return 25;
  }

  final today = referenceDate ?? DateTime.now();
  var age = today.year - birthDate.year;
  final hadBirthday =
      today.month > birthDate.month ||
      (today.month == birthDate.month && today.day >= birthDate.day);
  if (!hadBirthday) {
    age -= 1;
  }

  return age.clamp(0, 150).toInt();
}

DateTime? _parseBirthDate(String? value) {
  final raw = value?.trim();
  if (raw == null || raw.isEmpty) {
    return null;
  }

  final separator = raw.contains('-')
      ? '-'
      : raw.contains('/')
      ? '/'
      : null;
  if (separator == null) {
    return null;
  }

  final parts = raw.split(separator);
  if (parts.length != 3) {
    return null;
  }

  final firstPart = int.tryParse(parts[0]);
  final secondPart = int.tryParse(parts[1]);
  final thirdPart = int.tryParse(parts[2]);
  if (firstPart == null || secondPart == null || thirdPart == null) {
    return null;
  }

  if (parts[0].length == 4) {
    return DateTime(firstPart, secondPart, thirdPart);
  }

  return DateTime(thirdPart, secondPart, firstPart);
}

double _toKilograms(double value, String unit) {
  return switch (unit.toLowerCase()) {
    'lb' => value * 0.45359237,
    'g' => value / 1000,
    _ => value,
  };
}

double _toCentimeters(double value, String unit) {
  return switch (unit.toLowerCase()) {
    'm' => value * 100,
    'ft' => value * 30.48,
    'in' => value * 2.54,
    _ => value,
  };
}

double _calculateBmi(double weightKg, double heightCm) {
  final heightM = heightCm / 100;
  if (heightM <= 0) {
    return 0;
  }

  return weightKg / (heightM * heightM);
}

String _normalizeObjective(String objective) {
  return switch (objective) {
    'loseWeight' => 'loseWeight',
    'gainMass' => 'gainMass',
    _ => 'maintainWeight',
  };
}

double _applyObjectiveAdjustment(
  double tdee,
  String objective,
  double bmi,
  String activityLevel,
) {
  if (objective == 'loseWeight') {
    final deficit = _resolveLossDeficitPercent(bmi, activityLevel);
    return tdee * (1 - deficit);
  }

  if (objective == 'gainMass') {
    final surplus = _resolveGainSurplusPercent(bmi, activityLevel);
    return tdee * (1 + surplus);
  }

  return tdee;
}

double _resolveLossDeficitPercent(double bmi, String activityLevel) {
  // Degraus originais: 30% (≥40), 27% (≥35), 25% (≥30), 22% (≥27), 18%.
  // Interpolados para a meta não saltar ao cruzar um limiar de IMC.
  var percent = 0.18;

  if (bmi >= 40) {
    percent = 0.3;
  } else if (bmi >= 35) {
    percent = _lerp(0.27, 0.3, (bmi - 35) / 5);
  } else if (bmi >= 30) {
    percent = _lerp(0.25, 0.27, (bmi - 30) / 5);
  } else if (bmi >= 27) {
    percent = _lerp(0.22, 0.25, (bmi - 27) / 3);
  } else if (bmi >= 22) {
    percent = _lerp(0.18, 0.22, (bmi - 22) / 5);
  }

  if (activityLevel == 'very' || activityLevel == 'extreme') {
    percent -= 0.02;
  }

  return percent.clamp(0.15, 0.3);
}

double _resolveGainSurplusPercent(double bmi, String activityLevel) {
  var percent = 0.1;

  if (bmi < 18.5) {
    percent = 0.2;
  } else if (bmi < 22) {
    percent = 0.15;
  } else if (bmi < 25) {
    percent = 0.12;
  }

  if (activityLevel == 'very' || activityLevel == 'extreme') {
    percent += 0.02;
  }

  return percent.clamp(0.08, 0.22);
}

int _resolveCalorieFloor(String sex) {
  return switch (sex.trim().toLowerCase()) {
    'feminino' => 1200,
    'masculino' => 1500,
    _ => 1350,
  };
}

double _resolveMetabolicWeightKg(
  double weightKg,
  double heightCm,
  String objective,
) {
  if (objective != 'loseWeight') {
    return weightKg;
  }

  final heightM = heightCm / 100;
  if (heightM <= 0) {
    return weightKg;
  }

  final bmi = weightKg / (heightM * heightM);
  final adjustedWeightKg = _resolveAdjustedWeightKg(weightKg, heightM);

  // IMC ≥ 30: peso ajustado (IBW + 25% do excesso) para não inflar o TDEE.
  // IMC ≤ 27: peso real. Entre 27 e 30 interpola — senão cruzar obesidade
  // (ex.: 87,8 → 86,2 kg / 1,70 m) sobe a meta de uma vez.
  if (bmi >= 30) {
    return adjustedWeightKg;
  }

  if (bmi <= 27) {
    return weightKg;
  }

  final fadeToActual = (30 - bmi) / 3;
  return _lerp(adjustedWeightKg, weightKg, fadeToActual);
}

double _resolveAdjustedWeightKg(double weightKg, double heightM) {
  final idealWeightKg = 24.9 * (heightM * heightM);
  final adjustedWeightKg = idealWeightKg + ((weightKg - idealWeightKg) * 0.25);
  if (adjustedWeightKg < idealWeightKg) {
    return idealWeightKg;
  }
  if (adjustedWeightKg > weightKg) {
    return weightKg;
  }
  return adjustedWeightKg;
}

double _lerp(double from, double to, double t) {
  final clamped = t.clamp(0.0, 1.0);
  return from + (to - from) * clamped;
}

_MacroGoals _resolveMacroGoals(
  int dailyCalories,
  String objective,
  double metabolicWeightKg,
) {
  final minProtein = (metabolicWeightKg * _resolveProteinPerKg(objective)).round();
  final minFat = (metabolicWeightKg * _resolveFatPerKg(objective)).round();

  final distribution = _resolveMacroDistribution(objective);
  var protein = ((dailyCalories * distribution.protein) / 4).round();
  var fat = ((dailyCalories * distribution.fat) / 9).round();

  protein = protein < minProtein ? minProtein : protein;
  fat = fat < minFat ? minFat : fat;

  var carbs = ((dailyCalories - (protein * 4) - (fat * 9)) / 4)
      .round()
      .clamp(0, 9999);

  final minCarbs =
      (metabolicWeightKg * _resolveCarbsPerKgFloor(objective)).round();
  if (carbs < minCarbs) {
    carbs = minCarbs;
  }

  final totalCalories = (protein * 4) + (fat * 9) + (carbs * 4);
  if (totalCalories > dailyCalories) {
    final overflow = totalCalories - dailyCalories;
    final fatReduction = ((overflow / 9).ceil()).clamp(0, fat - minFat);
    fat -= fatReduction;
  }

  final baseCalories = (protein * 4) + (fat * 9);
  carbs = ((dailyCalories - baseCalories) / 4).round().clamp(0, 9999);

  return _MacroGoals(protein: protein, fat: fat, carbs: carbs);
}

double _resolveProteinPerKg(String objective) {
  return switch (objective) {
    'loseWeight' => 2.2,
    'gainMass' => 1.8,
    _ => 1.6,
  };
}

double _resolveFatPerKg(String objective) {
  return objective == 'loseWeight' ? 0.7 : 0.8;
}

double _resolveCarbsPerKgFloor(String objective) {
  return switch (objective) {
    'loseWeight' => 1.5,
    'gainMass' => 2.5,
    _ => 2.0,
  };
}

_MacroDistribution _resolveMacroDistribution(String objective) {
  return switch (objective) {
    'loseWeight' => const _MacroDistribution(protein: 0.35, fat: 0.30),
    'gainMass' => const _MacroDistribution(protein: 0.25, fat: 0.25),
    _ => const _MacroDistribution(protein: 0.25, fat: 0.30),
  };
}

class _MacroDistribution {
  const _MacroDistribution({required this.protein, required this.fat});

  final double protein;
  final double fat;
}

class _MacroGoals {
  const _MacroGoals({
    required this.protein,
    required this.fat,
    required this.carbs,
  });

  final int protein;
  final int fat;
  final int carbs;
}

File: backend/src/auth/utils/nutrition-goal-calculator.ts
type Objective = 'loseWeight' | 'gainMass' | 'maintainWeight';

export type NutritionProfileInput = {
  birthDate?: string | null;
  weight?: number | string | null;
  height?: number | string | null;
  weightUnit?: string | null;
  heightUnit?: string | null;
  sex?: string | null;
  objective?: string | null;
  activityLevel?: string | null;
};

export type NutritionGoalOutput = {
  dailyCalorieGoal: number;
  dailyProteinGoal: number;
  dailyCarbsGoal: number;
  dailyFatGoal: number;
};

export function calculateNutritionGoalsFromProfile(
  input: NutritionProfileInput,
): NutritionGoalOutput | null {
  const weightKg = toKilograms(input.weight, input.weightUnit);
  const heightCm = toCentimeters(input.height, input.heightUnit);
  const age = calculateAgeFromBirthDate(input.birthDate);

  if (weightKg <= 0 || heightCm <= 0 || age <= 0) {
    return null;
  }

  const objective = normalizeObjective(input.objective);
  const bmi = calculateBmi(weightKg, heightCm);
  const metabolicWeightKg = resolveMetabolicWeightKg(weightKg, heightCm, objective);

  const sexFactor = resolveSexFactor(input.sex);
  const bmr = (10 * metabolicWeightKg) + (6.25 * heightCm) - (5 * age) + sexFactor;
  const activityMultiplier = resolveActivityMultiplier(input.activityLevel);
  const tdee = bmr * activityMultiplier;

  const adjustedCalories = applyObjectiveAdjustment(
    tdee,
    objective,
    bmi,
    input.activityLevel,
  );
  const minimumCalories = resolveCalorieFloor(input.sex);
  const dailyCalorieGoal = Math.max(minimumCalories, Math.round(adjustedCalories));

  const macros = resolveMacroGoals(
    dailyCalorieGoal,
    objective,
    metabolicWeightKg,
  );

  return {
    dailyCalorieGoal,
    dailyProteinGoal: macros.dailyProteinGoal,
    dailyCarbsGoal: macros.dailyCarbsGoal,
    dailyFatGoal: macros.dailyFatGoal,
  };
}

function calculateAgeFromBirthDate(value?: string | null): number {
  if (!value) {
    return 0;
  }

  const birthDate = new Date(value);
  if (Number.isNaN(birthDate.getTime())) {
    return 0;
  }

  const today = new Date();
  let age = today.getFullYear() - birthDate.getFullYear();
  const hadBirthday =
    today.getMonth() > birthDate.getMonth() ||
    (today.getMonth() === birthDate.getMonth() &&
      today.getDate() >= birthDate.getDate());

  if (!hadBirthday) {
    age -= 1;
  }

  return Math.max(0, Math.min(150, age));
}

function toKilograms(
  value: number | string | null | undefined,
  unit: string | null | undefined,
): number {
  const numeric = toNumber(value);
  if (numeric <= 0) {
    return 0;
  }

  switch ((unit ?? 'kg').toLowerCase()) {
    case 'lb':
      return numeric * 0.45359237;
    case 'g':
      return numeric / 1000;
    default:
      return numeric;
  }
}

function toCentimeters(
  value: number | string | null | undefined,
  unit: string | null | undefined,
): number {
  const numeric = toNumber(value);
  if (numeric <= 0) {
    return 0;
  }

  switch ((unit ?? 'cm').toLowerCase()) {
    case 'm':
      return numeric * 100;
    case 'ft':
      return numeric * 30.48;
    case 'in':
      return numeric * 2.54;
    default:
      return numeric;
  }
}

function calculateBmi(weightKg: number, heightCm: number): number {
  const heightM = heightCm / 100;
  if (heightM <= 0) {
    return 0;
  }

  return weightKg / (heightM * heightM);
}

function resolveActivityMultiplier(activityLevel?: string | null): number {
  switch ((activityLevel ?? 'sedentary').toLowerCase()) {
    case 'lightly':
      return 1.375;
    case 'moderate':
      return 1.55;
    case 'very':
      return 1.725;
    case 'extreme':
      return 1.9;
    default:
      return 1.2;
  }
}

function resolveSexFactor(sex?: string | null): number {
  switch ((sex ?? '').trim().toLowerCase()) {
    case 'feminino':
      return -161;
    case 'masculino':
      return 5;
    default:
      return -78;
  }
}

function resolveCalorieFloor(sex?: string | null): number {
  switch ((sex ?? '').trim().toLowerCase()) {
    case 'feminino':
      return 1200;
    case 'masculino':
      return 1500;
    default:
      return 1350;
  }
}

function normalizeObjective(objective?: string | null): Objective {
  switch (objective) {
    case 'loseWeight':
    case 'gainMass':
      return objective;
    default:
      return 'maintainWeight';
  }
}

function applyObjectiveAdjustment(
  tdee: number,
  objective: Objective,
  bmi: number,
  activityLevel?: string | null,
): number {
  if (objective === 'loseWeight') {
    const deficitPercent = resolveLossDeficitPercent(bmi, activityLevel);
    return tdee * (1 - deficitPercent);
  }

  if (objective === 'gainMass') {
    const surplusPercent = resolveGainSurplusPercent(bmi, activityLevel);
    return tdee * (1 + surplusPercent);
  }

  return tdee;
}

function resolveLossDeficitPercent(
  bmi: number,
  activityLevel?: string | null,
): number {
  let percent = 0.2;

  if (bmi >= 40) {
    percent = 0.3;
  } else if (bmi >= 35) {
    percent = 0.27;
  } else if (bmi >= 30) {
    percent = 0.25;
  } else if (bmi >= 27) {
    percent = 0.22;
  } else {
    percent = 0.18;
  }

  const normalizedActivity = (activityLevel ?? '').toLowerCase();
  if (normalizedActivity === 'very' || normalizedActivity === 'extreme') {
    percent -= 0.02;
  }

  return clamp(percent, 0.15, 0.3);
}

function resolveGainSurplusPercent(
  bmi: number,
  activityLevel?: string | null,
): number {
  let percent = 0.1;

  if (bmi < 18.5) {
    percent = 0.2;
  } else if (bmi < 22) {
    percent = 0.15;
  } else if (bmi < 25) {
    percent = 0.12;
  }

  const normalizedActivity = (activityLevel ?? '').toLowerCase();
  if (normalizedActivity === 'very' || normalizedActivity === 'extreme') {
    percent += 0.02;
  }

  return clamp(percent, 0.08, 0.22);
}

function resolveMetabolicWeightKg(
  weightKg: number,
  heightCm: number,
  objective: Objective,
): number {
  if (objective !== 'loseWeight') {
    return weightKg;
  }

  const heightM = heightCm / 100;
  if (heightM <= 0) {
    return weightKg;
  }

  const bmi = weightKg / (heightM * heightM);
  if (bmi < 30) {
    return weightKg;
  }

  const idealWeightKg = 24.9 * (heightM * heightM);
  const adjustedWeightKg = idealWeightKg + ((weightKg - idealWeightKg) * 0.25);
  return Math.max(idealWeightKg, Math.min(weightKg, adjustedWeightKg));
}

function resolveMacroGoals(
  dailyCalories: number,
  objective: Objective,
  metabolicWeightKg: number,
): {
  dailyProteinGoal: number;
  dailyFatGoal: number;
  dailyCarbsGoal: number;
} {
  const proteinPerKg = resolveProteinPerKg(objective);
  const fatPerKg = resolveFatPerKg(objective);

  const minProteinGrams = Math.round(metabolicWeightKg * proteinPerKg);
  const minFatGrams = Math.round(metabolicWeightKg * fatPerKg);

  const distribution = resolveMacroDistribution(objective);
  let proteinGrams = Math.round((dailyCalories * distribution.protein) / 4);
  let fatGrams = Math.round((dailyCalories * distribution.fat) / 9);

  proteinGrams = Math.max(proteinGrams, minProteinGrams);
  fatGrams = Math.max(fatGrams, minFatGrams);

  let carbsGrams = Math.max(
    0,
    Math.round((dailyCalories - (proteinGrams * 4) - (fatGrams * 9)) / 4),
  );

  const minCarbsGrams = Math.round(
    metabolicWeightKg * resolveCarbsPerKgFloor(objective),
  );

  if (carbsGrams < minCarbsGrams) {
    carbsGrams = minCarbsGrams;
  }

  const totalCalories = (proteinGrams * 4) + (fatGrams * 9) + (carbsGrams * 4);
  if (totalCalories > dailyCalories) {
    const overflow = totalCalories - dailyCalories;
    const fatReduction = Math.min(fatGrams - minFatGrams, Math.ceil(overflow / 9));
    fatGrams -= Math.max(0, fatReduction);
  }

  const recalculatedTotal = (proteinGrams * 4) + (fatGrams * 9);
  carbsGrams = Math.max(0, Math.round((dailyCalories - recalculatedTotal) / 4));

  return {
    dailyProteinGoal: proteinGrams,
    dailyFatGoal: fatGrams,
    dailyCarbsGoal: carbsGrams,
  };
}

function resolveProteinPerKg(objective: Objective): number {
  if (objective === 'loseWeight') {
    return 2.2;
  }

  if (objective === 'gainMass') {
    return 1.8;
  }

  return 1.6;
}

function resolveFatPerKg(objective: Objective): number {
  if (objective === 'loseWeight') {
    return 0.7;
  }

  return 0.8;
}

function resolveCarbsPerKgFloor(objective: Objective): number {
  if (objective === 'loseWeight') {
    return 1.5;
  }

  if (objective === 'gainMass') {
    return 2.5;
  }

  return 2;
}

function resolveMacroDistribution(objective: Objective): {
  protein: number;
  fat: number;
} {
  if (objective === 'loseWeight') {
    return { protein: 0.35, fat: 0.3 };
  }

  if (objective === 'gainMass') {
    return { protein: 0.25, fat: 0.25 };
  }

  return { protein: 0.25, fat: 0.3 };
}

function clamp(value: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, value));
}

function toNumber(value: number | string | null | undefined): number {
  if (typeof value === 'number') {
    return Number.isFinite(value) ? value : 0;
  }

  if (typeof value === 'string') {
    const normalized = value.replace(',', '.').trim();
    const parsed = Number(normalized);
    return Number.isFinite(parsed) ? parsed : 0;
  }

  return 0;
}
